File: frontend/lib/calculator.ts
import type { UserProfile, NutritionPlan } from '@/types'

export const ACTIVITY_LABELS: Record<string, string> = {
  sedentary: 'Сидячий (нет спорта)',
  light: 'Лёгкая активность (1–3 раза/нед)',
  moderate: 'Умеренная (3–5 раз/нед)',
  active: 'Активный (6–7 раз/нед)',
  very_active: 'Очень активный (2 раза в день)',
}

const ACTIVITY_COEFFICIENTS: Record<string, number> = {
  sedentary: 1.2,
  light: 1.375,
  moderate: 1.55,
  active: 1.725,
  very_active: 1.9,
}

const GOAL_DELTA: Record<string, number> = {
  loss: -500,
  maintain: 0,
  gain: 500,
}

export const GOAL_LABELS: Record<string, string> = {
  loss: 'Похудение (−500 ккал)',
  maintain: 'Поддержание веса',
  gain: 'Набор массы (+500 ккал)',
}

export function calculateBMR(profile: UserProfile): number {
  const base = 10 * profile.weight + 6.25 * profile.height - 5 * profile.age
  return profile.gender === 'male' ? base + 5 : base - 161
}

export function calculateTDEE(bmr: number, activity: string): number {
  return bmr * (ACTIVITY_COEFFICIENTS[activity] ?? 1.2)
}

export function applyGoal(tdee: number, goal: string): number {
  return tdee + (GOAL_DELTA[goal] ?? 0)
}

export function calculateMacros(calories: number): NutritionPlan {
  return {
    calories: Math.round(calories),
    protein: Math.round((calories * 0.3) / 4),
    fat: Math.round((calories * 0.25) / 9),
    carbs: Math.round((calories * 0.45) / 4),
  }
}

export function calculateNutrition(profile: UserProfile): NutritionPlan & { bmr: number; tdee: number } {
  const bmr = calculateBMR(profile)
  const tdee = calculateTDEE(bmr, profile.activity)
  const calories = applyGoal(tdee, profile.goal)
  const macros = calculateMacros(calories)
  return { ...macros, bmr: Math.round(bmr), tdee: Math.round(tdee) }
}

File: modules/calculator.py
"""
Модуль расчёта суточной нормы калорий и БЖУ.
Формула BMR: Mifflin-St Jeor (1990) — самая точная для большинства людей.
"""
from models.schemas import UserProfile, NutritionPlan


# Коэффициенты активности для расчёта TDEE = BMR × коэффициент
ACTIVITY_COEFFICIENTS: dict[str, float] = {
    "Сидячий (нет спорта)": 1.2,
    "Лёгкая активность (1–3 раза/нед)": 1.375,
    "Умеренная активность (3–5 раз/нед)": 1.55,
    "Активный (6–7 раз/нед)": 1.725,
    "Очень активный (2 раза в день)": 1.9,
}

# Корректировка калорий по цели
GOAL_DELTA: dict[str, float] = {
    "Похудение": -500.0,
    "Поддержание веса": 0.0,
    "Набор массы": +500.0,
}


def calculate_bmr(profile: UserProfile) -> float:
    """
    Рассчитывает базовый обмен веществ (BMR) по формуле Mifflin-St Jeor.

    Мужчины: 10*вес + 6.25*рост - 5*возраст + 5
    Женщины: 10*вес + 6.25*рост - 5*возраст - 161
    """
    base = 10 * profile.weight + 6.25 * profile.height - 5 * profile.age
    if profile.gender == "Мужской":
        return base + 5
    else:
        return base - 161


def calculate_tdee(bmr: float, activity: str) -> float:
    """Рассчитывает суточный расход энергии (TDEE) = BMR × коэффициент активности."""
    coefficient = ACTIVITY_COEFFICIENTS.get(activity, 1.2)
    return bmr * coefficient


def apply_goal(tdee: float, goal: str) -> float:
    """Корректирует калории в зависимости от цели (±500 ккал или без изменений)."""
    delta = GOAL_DELTA.get(goal, 0.0)
    return tdee + delta


def calculate_macros(calories: float) -> NutritionPlan:
    """
    Рассчитывает БЖУ из итоговых калорий по соотношению 30% / 25% / 45%.

    Белки и углеводы: 4 ккал/г
    Жиры: 9 ккал/г
    """
    protein = round((calories * 0.30) / 4, 1)
    fat = round((calories * 0.25) / 9, 1)
    carbs = round((calories * 0.45) / 4, 1)
    return NutritionPlan(
        calories=round(calories, 1),
        protein=protein,
        fat=fat,
        carbs=carbs,
    )


def calculate_nutrition(profile: UserProfile) -> NutritionPlan:
    """Полный расчёт: BMR → TDEE → цель → БЖУ. Главная функция модуля."""
    bmr = calculate_bmr(profile)
    tdee = calculate_tdee(bmr, profile.activity)
    target_calories = apply_goal(tdee, profile.goal)
    return calculate_macros(target_calories)
